jax/_src/nn/functions.py

```diff
@@ -765,3 +767,174 @@ def hard_silu(x: ArrayLike) -> Array:
+
+def _get_large_negative(dtype):
+  dtype_max = jnp.finfo(dtype).max
+  return jnp.asarray(-0.7 * dtype_max, dtype=dtype)
+
+def _get_causal_mask(T, dtype):
+  large_negative_number = _get_large_negative(dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+def _dot_product_attention_xla(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None,
+    mask: ArrayLike | None,
+    is_causal: bool,
+    scale: float):
+  # Compute the attention logits
+  logits = jnp.einsum('BTNH,BSNH->BNTS', query, key)
+
+  # Bias
+  if bias is not None:
+    logits = logits + bias.astype(logits.dtype)
+
+  # Logits scaling
+  scale = jnp.array(scale, dtype=logits.dtype)
+  logits = jnp.multiply(logits, scale)
+
+  # Mask
+  if mask is not None:
+    if mask.dtype == jnp.bool_:
+      large_negative_number = _get_large_negative(logits.dtype)
+      padded_logits = jnp.where(mask, logits, large_negative_number)
+    else:
+      padded_logits = logits + mask
+  else:
+    padded_logits = logits
+
+  if is_causal:
+    T, S = query.shape[-3], key.shape[-3]
+    assert T == S, f'T should equal to S, but got {T} vs {S}'
+    mask = _get_causal_mask(T, logits.dtype)
+    padded_logits = padded_logits + mask
+
+  # Softmax and it is always carried out in fp32.
+  padded_logits = padded_logits.astype(jnp.float32)
+  probs = jax.nn.softmax(padded_logits, axis=-1).astype(key.dtype)
+
+  # Compute the attention context.
+  encoded = jnp.einsum('BNTS,BSNH->BTNH', probs, value)
+  return encoded, probs
+
+def dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: str | None = None) -> Array:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. Two types of masks are
```

Comment:
what does the float mask buy us here? Do we need it?

Reply:
I think it is a general practice to use float masks to do the additive masking. Like in praxis, the mask is already using floats of large negative values ([here](https://github.com/google/praxis/blob/cdc67379f8704344b1f9fe6a0264222a54c98949/praxis/layers/attentions.py#L1678-L1679)). Also, the pytorch sdpa ([here](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)) also supports both. So, I think maybe we should do the same.

Reply:
Can't you do the same with the bias term, which pytorch doesn't have?

Reply:
But also: what does this give you that you can't do with the boolean mask?

Reply:
If there is any question: its easier to add features than remove features. Can always add this to a later PR.

Reply:
Yep, I think it's a good point. We can let user to pass anything to `bias` that needs to be added to the logits. So, the float mask should be passed into `bias`, or,  if users have the relative position bias, they should do the addition of `relative_bias+float_mask` and pass it to the `bias`. 

Then the `mask` is only for the boolean mask as you suggested. This simplified our design. Do I get it correct? I will update.

Reply:
Done